### PR TITLE
Tidy up readme

### PR DESCRIPTION
**Pull request by NicMcPhee**
Fix various issues in the README, like spelling mistakes, and add the BetterCodeHub badge.

### CODE REVIEW

**Comment by floogulinc** [comment]
Why switch everything from `-` to `*`?

**Comment by NicMcPhee** [comment]
> Why switch everything from `-` to `*`?

I (personally) prefer `*` to `-` for lists and Markdown lint wants all the list item markers to be the same throughout a file (which is sorta reasonable, I guess). If I don't have them all be the same, then my VS Code Markdown extension clutters up my world with a bazillion alerts. Since I was going through the whole README anyway, it wasn't a big deal to fix all those as I went.

**Comment by NicMcPhee** [comment]
As I mentioned in Slack and in the last commit:

> I'm seriously considering modifying the end game to use pull requests instead of hand merging. GitHub's support for PRs has gotten fancier and niftier, including warnings when your branch is behind `main` and giving you a push-button way to merge `main` into your PR branch.
>
> We definitely want them to be using pull requests and code reviews through the semester, so maybe we should just start with that.
>
> That will probably slow things down, but I'm thinking that this is going to become a two lab-session event in a purely on-line COVID world this semester anyway.

Thoughts on that?

File: README.md [hunk 1]
```diff
@@ -390,17 +428,20 @@ This tells you that you've successfully staged `user_info/Pat_and_Chris.md`.
...
-**Commit the change** _to your local copy of the repository_ with `git commit`
-on the command line.
+**Commit the change** _to your local copy of the repository_ with
 
-- This will bring up a (command line) editor for you to enter
+```bash
+git commit
+```
+
+* This will bring up a (command line) editor for you to enter
```
Comment (floogulinc) [inline README.md:437]: The editor may not be a command line one. For students on Windows we have even suggested they change the setting when installing Git to VS Code:

image

(https://gist.github.com/floogulinc/2d9f80b278332b0df731e777c43b206c#install-git)


Though vim is the default when you run through the installer and we could update the setup guide to suggest this. 

image

File: README.md [hunk 2]
```diff
@@ -145,14 +165,15 @@ local changes up to the GitHub repository so that other people can access them.
...
 account](https://help.github.com/articles/setting-your-commit-email-address-in-git/).
-You only need to do this once and it will "stick" throughout the lab
-in all systems that use `git`.
+You only need to do this once and it will "stick" throughout the course
+(and beyond) in all systems that use `git`.
 
-- Open a terminal window
-- Type `git config --global user.email "email@example.com"` where
+* Open a terminal window
+  * :mask: On Macs and Linux boxes use the Terminal app. On Windows use Powershell.
```
Comment (floogulinc) [inline README.md:172]:
```suggestion
  * :mask: On Macs and Linux boxes use the Terminal app. On Windows use PowerShell.
```

**Review by floogulinc — CHANGES_REQUESTED** [review]

**Comment by floogulinc** [comment]
@NicMcPhee Do you think we should have them use token authentication in this lab? It will be required by GitHub starting in August: https://github.blog/2020-12-15-token-authentication-requirements-for-git-operations/

File: README.md [hunk 3]
```diff
@@ -531,28 +572,43 @@ tests to your feature branch, then it will have no impact on people
...
-development branches that are merged into `master` when that work is deemed
+development branches that are merged into `main` when that work is deemed
 "done done" (passes the tests, has been through a code review, etc.).
 
-To illustrate the use of branches, we'll have each group create a new branch
+> Historically the default has actually been called `master`. One side-effect
+> of the Black Lives Matter movement has been an increasing recognition of
+> the painful connotations that term has for many people, and there have
+> been [calls to change from `master` to the more neutral `main`](https://dev.to/afrodevgirl/replacing-master-with-main-in-github-2fjf).
+> GitHub is in fact working to change the default naming for new projects,
```
[the remark this answers is not in the document]
Comment (NicMcPhee) [reply README.md:585]: I assume that copies made from templates (what GitHub Classroom does) will follow the existing branch names (it really has to), so we still have to change our branch names in our repos.

I actually meant to link to the official page, and obviously gapped it – thanks for catching that.

File: README.md (hunk 1 above)
Comment (NicMcPhee) [reply README.md:437]: Thanks for that. I'm fine with them using VS Code as their editor, so I'll just update my text.

**Comment by NicMcPhee** [comment]
We probably should introduce token authentication here.

@floogulinc Do you think we should split this into two distinct labs, possibly run on adjacent lab sessions (one on Tuesday and one on Thursday)? The write-up has gotten really long, and adding token authentication and PRs is going to make it longer.

Maybe do the first part (editing text files with little change of merge conflicts) on the first day, and then the second part (editing the code and dealing with PRs and merge conflicts) on the second day.

It's going to screw with the calendar, but with COVID and everything on-line the calendar is going to have to change/be flexible anyway.

**Comment by NicMcPhee** [comment]
Actually, two _labs_ wouldn't be a good idea because then we'd have two repos and lose what we'd done in one when we start the other. So maybe split the write-up into two Markdown documents, with the first linking (at the end) to the second? Or maybe we just want one really long document. I'm not sure.

**Comment by NicMcPhee** [comment]
@floogulinc I hadn't actually gotten into the personal access token thing yet, and the one part that looks a little intimidating is the scopes thing. The "simple" thing would be to give myself "all the scopes" with my token, but I'm guessing that's not "best practice". On the other hand, having and managing a bazillion scopes seems like a Big Mess Waiting To Happen. How are you handling this yourself?

**Comment by NicMcPhee** [comment]
@floogulinc Adding personal access tokens, pull requests, etc., to this would be substantial changes in their own right. I'm kind of thinking those should be in separate issues/PRs. Does that make sense to you? If so, I think we could call this good and merge it in, and then I could think about doing the other pieces separately.

**Comment by NicMcPhee** [comment]
Commit 2bb64d3 closes #55

File: .vscode/settings.json [hunk 4]
```diff
@@ -1,5 +1,6 @@
 {
     "java.configuration.updateBuildConfiguration": "automatic",
+    "java.server.launchMode": "Standard"
```
Comment (floogulinc) [inline .vscode/settings.json:3]:
```suggestion
    "java.server.launchMode": "Standard",
```

File: README.md [hunk 5]
```diff
@@ -76,20 +86,43 @@ because there are moments where we're expecting certain things to
...
+  docs. There are a _lot_ more than just these, but these are short and pretty easy to spin through, and cover the key concepts we’ll need at the beginning of the semester.
+  * [GitHub Hello World](https://guides.github.com/activities/hello-world/) (a really nice,
+    short introduction to the key ideas)
+  * [Getting started with GitHub](https://docs.github.com/en/free-pro-team@latest/github/getting-started-with-github)
+  * [Understanding the GitHub flow](https://guides.github.com/introduction/flow/)
+  * [Mastering Markdown](https://guides.github.com/features/mastering-markdown/)
+    (Markdown is the markup language used
+    by Slack, GitHub, and _many_ other
+    online tools, so understanding at least
+    the basics is really valuable.)
```
Comment (floogulinc) [inline README.md:109]: Should this be on a sub-bullet like the one below is?

File: README.md [hunk 6]
```diff
@@ -162,14 +195,40 @@ local changes up to the GitHub repository so that other people can access them.
 
 ### Configuring your `git` e-mail
 
+:mask::warning::exclamation::warning::mask:
+This whole section may be irrelevant for home set ups. It looks like
```
Comment (floogulinc) [inline README.md:199]:
```suggestion
This whole section may be irrelevant for home set-ups. It looks like
```

File: README.md [hunk 7]
```diff
@@ -312,37 +371,37 @@ install them one at a time if you want to know more about what we're doing here.
...
+
+As well as running tests from an "external" terminal window, you can also run
+them from a terminal inside VS Code. Go to the "Terminal" menu and select
+"New Terminal". That should bring up a terminal along the bottom of your VS Code
+window. That terminal is just like an "external" terminal and you can use it
+to do anything you could have done in an "external" terminal window. This
+includes things like:
+
+```bash
+   ./gradlew test
```
Comment (floogulinc) [inline README.md:392]:
```suggestion
./gradlew test
```

**Review by floogulinc — CHANGES_REQUESTED** [review]

File: README.md (hunk 7 above)
Comment (NicMcPhee) [reply README.md:400]: I actually indented it on purpose as I think it reads a little better indented. I guess is could be a source of confusion, but that doesn't seem a huge risk?

File: README.md (hunk 5 above)
Comment (NicMcPhee) [reply README.md:117]: Yup, it almost certainly should. Thanks!

File: README.md [hunk 8]
```diff
@@ -660,232 +847,374 @@ commit and push your work.
...
+There are numerous ways this conflict could have come about, such as:
+
+* Chris wants to change the line from calling `patSaysHello()` to
+  instead called the (supposedly different) method `chrisSaysHello()`.
+* Chris has renamed `patSaysHello()` to `chrisSaysHello()`, and wants to change
+  the call to match this renaming.
+* Chris wants to make _both_ function calls (in some order, which `git` couldn't
+  begin to guess).
+
+Here `git can't tell which of these happened, and thus can't "know" how to
```
Comment (helloworld12321) [inline README.md:1103]:
```suggestion
Here `git` can't tell which of these happened, and thus can't "know" how to
```

Looks like you're missing an accent, by the way! :)

File: README.md [hunk 9]
```diff
@@ -660,232 +847,374 @@ commit and push your work.
...
+handle the merge. Thus it punts it to us, the human users to sort out. Sometimes
```
Comment (helloworld12321) [inline README.md:1104]:
```suggestion
handle the merge. Thus it punts it to us, the human users, to sort out. Sometimes
```
(also a comma)

File: README.md [hunk 10]
```diff
@@ -660,232 +847,374 @@ commit and push your work.
...
+that the tests are failing because they have the alphabetical order wrong,
+they'll need to pull down these changes, fix the alphabetical order, push back
+up, and attempt to merge again.
+
+##### You need a review
+
+Now you've finally gotten your code up-to-date and all the checks pass. You
+still can't merge because of a big red X and the message "Review required".
+This is because we're following a common "best practice" and requiring at
+least one code review _by another person_ for each pull request.
```
Comment (floogulinc) [inline README.md:1135]: Are we changing this to two code reviews? If so:
```suggestion
least two code review _by other people_ for each pull request.
```

**Review by floogulinc — CHANGES_REQUESTED** [review]

File: README.md [hunk 11]
```diff
@@ -660,232 +847,374 @@ commit and push your work.
...
+
+So you need to get someone to review your code! In the upper right there's
+a "Reviewers" section. Click on the gear there and you should get a drop down
+with the (GitHub) names of the other people in the class. Choose two or three
+and click their names to invite them to review your code. _Don't just choose
+the top three – if everyone does that the same few people will be asked to do
+_all_ the reviews. Instead, try choosing the two or three names right after yours
+on the list – that should spread the load around.
```
Comment (helloworld12321) [inline README.md:1143]: It looks like this `_` is unclosed. I can't tell what your intention was here, but as a suggestion, maybe we want to leave this sentence un-italicized?

```suggestion
and click their names to invite them to review your code. Don't just choose
the top three – if everyone does that the same few people will be asked to do
_all_ the reviews. Instead, try choosing the two or three names right after yours
on the list – that should spread the load around.
```

File: INFO_ON_MERGE_CONFLICTS.md [hunk 12]
```diff
@@ -0,0 +1,163 @@
+# Info on merge conflicts <!-- omit in toc -->
```
Comment (floogulinc) [inline INFO_ON_MERGE_CONFLICTS.md:1]: What do you think about just calling the file `MERGE_CONFLICTS.MD`?

File: INFO_ON_MERGE_CONFLICTS.md [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+
+Working as part of a team on a shared repository can be tricky.
+If no one else was working on the repository life would be easy, and
```
Comment (floogulinc) [inline INFO_ON_MERGE_CONFLICTS.md:4]:
```suggestion
If no one else was working on the repository, life would be easy, and
```

**Review by floogulinc — DISMISSED** [review]
Having read through most of the Merge Conflicts doc, I think we should rewrite it to better reflect a workflow that uses pull requests. We can use examples of either:
- A student makes a PR but something is later merged into main and they need to update their branch _from_ main and opt to do it and deal with the merge conflict in GitKraken rather than the GitHub UI (one can be forced to do this if its too complicated for GitHub).
- Two students are working on the same feature branch asynchronously and as such sometimes need to resolve conflicts on the same branch.

My main concern with how this is now is with branch protection on student repos, they will probably never be merging things into `main`, rather merging `main` into their feature branches to keep them up to date prior to merging the pull request in GitHub.

File: README.md (hunk 7 above)
Comment (floogulinc) [reply README.md:400]: It does seem odd for that one to be different from all the other instances in the doc.

File: README.md (hunk 10 above)
Comment (NicMcPhee) [reply README.md:1135]: Good catch – thanks!

**Comment by NicMcPhee** [comment]
I agree with @floogulinc that the current merge conflicts file really isn't "right" and doesn't reflect the kind of workflow we're expecting these days. Unfortunately, I don't think I have time to fix that now. The question then is whether to leave it in as is, or take it out altogether.

Thanks a ton to both @helloworld12321 and @floogulinc for all the excellent copy editing. If we never added new (broken) text, this thing would just get closer to perfect every semester! 😜

File: INFO_ON_MERGE_CONFLICTS.md (hunk 12 above)
Comment (NicMcPhee) [reply INFO_ON_MERGE_CONFLICTS.md:1]: Done – thanks!

**Comment by NicMcPhee** [comment]
I created a new issue #71 to capture need to eventually update `MERGE_CONFLICTS.md` to better reflect the desired pull-request-based workflow. That's not happening for this week, though, although folks are more than welcome to work on it in the meantime so it's ready when @kklamberty teaches the class in the Fall. 😀

**Pull request merged**